discovery-provider/src/tasks/index_trending.py

```diff
@@ -209,6 +212,114 @@ def index_trending(self, db: SessionManager, redis: Redis, timestamp):
 
+    index_trending_notifications(db, timestamp)
+
+
+def index_trending_notifications(db: SessionManager, timestamp: datetime):
+    # Get the top 5 trending tracks from the new trending calculations
+    # Get the most recent trending tracks notifications
+    # Calculate any diff and write the new notifications if the trending track has moved up in rank
+    # Skip if the user was notified of the trending track within the last TRENDING_INTERVAL_HOURS
+    # Skip If the new rank is not less than the old rank, skip
+    #   ie. Skip if track moved from #2 trending to #3 trending or stayed the same
+    trending_strategy_factory = TrendingStrategyFactory()
+    # The number of tracks to notify for in the top
+    NOTIFICATIONS_TRACK_LIMIT = 5
+    with db.scoped_session() as session:
+        trending_tracks = _get_trending_tracks_with_session(
+            session,
+            {time: "week"},
+            trending_strategy_factory.get_strategy(TrendingType.TRACKS),
+        )
+        top_trending = trending_tracks[:NOTIFICATIONS_TRACK_LIMIT]
+        top_trending_track_ids = [str(t["track_id"]) for t in top_trending]
+
+        previous_trending_notifications = (
+            session.query(Notification)
+            .filter(
+                Notification.type == "trending",
+                Notification.specifier.in_(top_trending_track_ids),
+            )
+            .all()
+        )
+
+        latest_notification_query = text(
+            """
+                SELECT 
+                    DISTINCT specifier,
+                    timestamp,
+                    data
+                FROM notification
+                WHERE 
+                    type=:type AND
+                    specifier in :track_ids
+                ORDER BY
+                    timestamp desc
+            """
+        )
+        latest_notification_query = latest_notification_query.bindparams(
+            bindparam("track_ids", expanding=True)
+        )
+
+        previous_trending_notifications = session.execute(
+            latest_notification_query,
+            {"track_ids": top_trending_track_ids, "type": "trending"},
+        )
+        previous_trending = {
+            n[0]: {"timestamp": n[1], **n[2]} for n in previous_trending_notifications
+        }
+        logger.info(previous_trending)
+
+        notifications = []
+
+        # Do not send notifications for the same track trending within 24 hours
+        NOTIFICATION_INTERVAL_SEC = 60 * 60 * 24
+
+        for index, track in enumerate(top_trending):
+            track_id = track["track_id"]
+            rank = index + 1
+            previous_track_notification = previous_trending.get(str(track["track_id"]))
+            logger.info(previous_track_notification)
+            if previous_track_notification is not None:
+                prev_notification_datetime = datetime.fromtimestamp(
+                    previous_track_notification["timestamp"].timestamp()
+                )
+                if (
```

Comment:
Are these checks possible to do in a sql query?

Reply:
Yes, it definitely is possible to contain all this logic in sql, but I think it would be really hard to read and understand. I don't think the tradeoff is worth it, since the amount of application code is minimal and readable.